cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationNotificationSubscriberService.java

```diff
@@ -0,0 +1,353 @@
+      throws IOException, TableNotFoundException {
+    return getAppMetadataStore(context).retrieveSubscriberState(getTopicId().getTopic(),
+        "operation.state.reader");
+  }
+
+  @Override
+  protected void storeMessageId(StructuredTableContext context, String messageId)
+      throws IOException, TableNotFoundException {
+    getAppMetadataStore(context).persistSubscriberState(getTopicId().getTopic(),
+        "operation.state.reader", messageId);
+  }
+
+  @Override
+  protected void processMessages(
+      StructuredTableContext structuredTableContext,
+      Iterator<ImmutablePair<String, Notification>> messages)
+      throws Exception {
+    List<Runnable> tasks = new LinkedList<>();
+    while (messages.hasNext()) {
+      ImmutablePair<String, Notification> messagePair = messages.next();
+      List<Runnable> runnables =
+          processNotification(
+              messagePair.getFirst().getBytes(StandardCharsets.UTF_8),
+              messagePair.getSecond(),
+              structuredTableContext);
+      tasks.addAll(runnables);
+    }
+
+    // Only add post-processing tasks if all messages are processed. If there is exception in the
+    // processNotification, messages will be replayed.
+    this.tasks.addAll(tasks);
+  }
+
+  @Override
+  protected void postProcess() {
+    Runnable task = tasks.poll();
+    while (task != null) {
+      task.run();
+      task = tasks.poll();
+    }
+  }
+
+  /**
+   * Process a {@link Notification} received from TMS.
+   *
+   * @param messageIdBytes the raw message id in the TMS for the notification
+   * @param notification the {@link Notification} to process
+   * @param context context to get the table for operations
+   * @return a {@link List} of {@link Runnable} tasks to run after the transactional processing of
+   *     the whole messages batch is completed
+   * @throws Exception if failed to process the given notification
+   */
+  @VisibleForTesting
+  List<Runnable> processNotification(
+      byte[] messageIdBytes,
+      Notification notification,
+      StructuredTableContext context)
+      throws Exception {
+    OperationRunStore runStore = new OperationRunStore(context);
+    List<Runnable> result = new ArrayList<>();
+    OperationNotification operationNotification;
+
+    try {
+      operationNotification = OperationNotification.fromNotification(
+          notification);
+    } catch (IllegalArgumentException | JsonSyntaxException e) {
+      LOG.warn("Got invalid operation notification", e);
+      return result;
+    }
+
+    handleOperationEvent(
+        operationNotification,
+        messageIdBytes,
+        runStore,
+        result
+    );
+
+    return result;
+  }
+
+  private void handleOperationEvent(
+      OperationNotification notification,
+      byte[] messageIdBytes,
+      OperationRunStore runStore,
+      List<Runnable> runnables
+  ) throws Exception {
+    LOG.trace("Processing operation status notification: {}", notification);
```

Comment:
This can probably be debug. If somebody hits an issue, it will be really hard to investigate after the fact without any log messages.